src/BuiltInTools/dotnet-watch/HotReload/AspNetCoreDeltaApplier.cs

```diff
@@ -0,0 +1,160 @@
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ExternalAccess.DotNetWatch;
+using Microsoft.Extensions.Tools.Internal;
+
+namespace Microsoft.DotNet.Watcher.Tools
+{
+    internal class AspNetCoreDeltaApplier : IDeltaApplier
+    {
+        private readonly IReporter _reporter;
+        private Task _task;
+        private NamedPipeServerStream _pipe;
+
+        public AspNetCoreDeltaApplier(IReporter reporter)
+        {
+            _reporter = reporter;
+        }
+
+        public async ValueTask InitializeAsync(DotNetWatchContext context, CancellationToken cancellationToken)
+        {
+            if (_pipe is not null)
+            {
+                _pipe.Close();
+                await _pipe.DisposeAsync();
+            }
+
+            _pipe = new NamedPipeServerStream("netcore-hot-reload", PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.Asynchronous | PipeOptions.CurrentUserOnly);
+            _task = _pipe.WaitForConnectionAsync(cancellationToken);
+
+            if (context.Iteration == 0)
+            {
+                var deltaApplier = Path.Combine(AppContext.BaseDirectory, "hotreload", "Microsoft.Extensions.AspNetCoreDeltaApplier.dll");
+                context.ProcessSpec.EnvironmentVariables.DotNetStartupHooks.Add(deltaApplier);
+
+                // Configure the app for EnC
+                context.ProcessSpec.EnvironmentVariables["DOTNET_MODIFIABLE_ASSEMBLIES"] = "debug";
+            }
+        }
+
+        public async ValueTask<bool> Apply(DotNetWatchContext context, string changedFile, DotNetWatchManagedModuleUpdatesWrapper? updates, CancellationToken cancellationToken)
+        {
+            if (!_task.IsCompletedSuccessfully || !_pipe.IsConnected)
+            {
+                // The client isn't listening
+                _reporter.Verbose("No client connected to receive delta updates.");
+                return false;
+            }
+
+            if (updates is null)
+            {
+                await context.BrowserRefreshServer.SendJsonSerlialized(new HotReloadApplied());
+                return true;
+            }
+
+            var payload = new UpdatePayload
+            {
+                ChangedFile = changedFile,
+                Deltas = updates.Value.Updates.Select(c => new UpdateDelta
+                {
+                    ModuleId = c.Module,
+                    ILDelta = c.ILDelta.ToArray(),
+                    MetadataDelta = c.MetadataDelta.ToArray(),
+                    UpdatedMethods = c.UpdatedMethods.ToArray(),
+                }),
+            };
+
+            // Jank mode. We should send this in a better (not json) format
+            await JsonSerializer.SerializeAsync(_pipe, payload, cancellationToken: cancellationToken);
+            await _pipe.FlushAsync(cancellationToken);
+
+            var result = ApplyResult.Failed;
+            var bytes = ArrayPool<byte>.Shared.Rent(1);
```

Comment:
Seems unnecessary to rent an array of size 1 here. I'd bet the array returned will be size 4096. If only one call to Apply is being run at a time, you could have it in a field.